### DISCUSSION TITLE
JSON for Sonoff 4CH irrigation valves

### DISCUSSION THREAD

**Discussion by jlg89** [RavenSystem/esp-homekit-devices#1864] [MEPLHAA Scripts Collection]
Device Name: Sonoff 4CHR3
URL to Device: https://itead.cc/product/sonoff-4ch-r3-pro-r3/
Used HAA version: 10.8.3
Summary of functions: Creates four sprinkler valve accessories

MEPLHAA script:
```
{
  "c": { "l": 13, "b": [{ "g": 0, "t": 5 }] },
  "a": [{
    "t": 20,
    "0": { "r": [{ "g": 12, "v": 0 }] },
    "1": { "r": [{ "g": 12, "v": 1 }] },
    "b": [{ "g": 0, "t": 1 }],
    "s": 0,
    "w": 1,
    "d": 1200
  },
  {
    "t": 20,
    "0": { "r": [{ "g": 5, "v": 0 }] },
    "1": { "r": [{ "g": 5, "v": 1 }] },
    "b": [{ "g": 9, "t": 1 }],
    "s": 0,
    "w": 1,
    "d": 1200
  },
  {
    "t": 20,
    "0": { "r": [{ "g": 4, "v": 0 }] },
    "1": { "r": [{ "g": 4, "v": 1 }] },
    "b": [{ "g": 10, "t": 1 }],
    "s": 0,
    "w": 1,
    "d": 1200
  },
  {
    "t": 20,
    "0": { "r": [{ "g": 15, "v": 0 }] },
    "1": { "r": [{ "g": 15, "v": 1 }] },
    "b": [{ "g": 14, "t": 1 }],
    "s": 0,
    "w": 1,
    "d": 1200
  }]
}
```
Long description:
Each valve can be controlled as a HomeKit accessory, with manual control via the buttons on the unit. One option is to link each valve to a zone switch in https://github.com/MTry/homebridge-smart-irrigation , though this may require a third-party app (e.g. https://apps.apple.com/us/app/controller-for-homekit/id1198176727) to configure the necessary Automations, as the Home app does not "see" the zone switches in that plugin as discrete accessories for use as Automation triggers.

Note that this JSON is specifically for the 4CHR3, the 4CHPRO may be different. The 4CHR3 is AC-only, and in order to use the relays to switch irrigation valves, a couple of traces must be cut on the PCB to disconnect the relay commons from the AC line. I'll be glad to post details if anyone is interested.

**Comment by aplok** [comment]
I am interested. Can you help me with automations ?

**Comment by jlg89** [comment]
You have to use the Controller app (https://apps.apple.com/us/app/controller-for-homekit/id1198176727) to do the automations -- at least the way I did them. If you're using homebridge-smart-irrigation, it has "dummy" zone control switches, so at a minimum you have to create automations that make the Sonoff switches match the corresponding dummy switches. Most irrigation controllers also use a master valve -- I call it "Valve 0" -- that must be turned on in order to get water to any of the watering zones, and should be turned off once watering is done. I used an additional Sonoff SV to control that master valve. Also, in order to prolong valve life, it's a good idea to "stagger" valve open/close functions to minimize pressure swings in the lines. A typical valve control sequence for a 4-station system might look something like this:

1. Zone 1 dummy switch goes on
2. Automation sees Zone 1 on, turns on Valve 0 and Valve 1
3. Zone 1 dummy switch goes off, Zone 2 dummy switch goes on
4. Automation sees Zone 1 off, activates a "delay-off" that turns off Valve 1 after 4s
5. Automation sees Zone 2 on, turns on Valve 2
6. etc.

Keeping Valve 1 on for 4s, while Valve 2 is opening, reduces the pressure on both valves and may increase valve life.

The automations I use:

Almost all of these must be created in Controller, as the standard Home app doesn't have the ability to create multi-conditional automations, and can't "see" the Zone switches in the irrigation plugin:

Valve 0 On -- if any Zone dummy switch is on, turn on Valve 0
Valve 0 Off -- if any Zone dummy switch is off AND all Zone dummy switches are off, turn off Valve 0 (sounds weird, but that's how it works)

Create these three for each zone/valve combo:
Zone 1 On -- if Zone 1 dummy switch goes on, turn on Valve 1
Valve 1 Off Delay -- if Zone 1 dummy switch goes off, turn on this switch (this is a switch in homebridge-dummy that automatically turns off after 4s)
Valve 1 Off -- if Valve 1 Off Delay turns off, turn off Valve 1 (this automation can be done in the Home app)

With all of that said, I want to state that, given some bugs in the homebridge-smart-irrigation plugin that aren't being addressed, and some overall performance concerns I have with the plugin's watering logic, I decided to bite the bullet and purchase a Rachio controller with a Tempest WeatherFlow gizmo. I've had great success with that setup at another location, and I'm tired of futzing with the plugin. So now I'll be looking for something else creative to do with the Sonoff 4CH...
